File: shell.c
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <spawn.h>
#include <sys/wait.h>
#include <unistd.h>
#include <errno.h>
#include <stdlib.h>

#define CMD "./new"
#define MAX_BUF 1024

extern char **environ;

int shellout(char *cmd);

int main() {
    struct timespec start, end;
    int i = 0, total_chars = 0;

    clock_gettime(CLOCK_MONOTONIC, &start);

    while (1) {
        int chars = shellout(CMD);
        if (chars < 0) {
            printf("Failed to run command '%s'\n", CMD);
            perror("shellout");
            return 127;
        }
        total_chars += chars;
        i++;

        clock_gettime(CLOCK_MONOTONIC, &end);
        if ((end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1E9 >= 1.0) {
            break;
        }
    }

    double cpu_time_used = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1E9;
    printf("C SHELL: Ran %d times in %f seconds.\n", i, cpu_time_used);

    return 0;
}

int shellout(char *cmd) {
    int pipefd[2];
    if (pipe(pipefd) == -1) {
        return -1;
    }

    pid_t pid;
    posix_spawn_file_actions_t action;
    posix_spawn_file_actions_init(&action);
    posix_spawn_file_actions_addclose(&action, pipefd[0]);  // close read end
    posix_spawn_file_actions_adddup2(&action, pipefd[1], STDOUT_FILENO);  // duplicate write end to stdout

    char *argv[] = {cmd, NULL};
    if (posix_spawn(&pid, cmd, &action, NULL, argv, environ) != 0) {
        return -1;
    }

    close(pipefd[1]);  // close write end

    char buffer[MAX_BUF];
    FILE *stream = fdopen(pipefd[0], "r");  // read from the reading end
    if (!stream) {
        return -1;
    }

    int total_chars = 0;
    while (fgets(buffer, sizeof(buffer), stream) != NULL) {
        total_chars += strlen(buffer);
    }

    fclose(stream);

    int status;
    waitpid(pid, &status, 0);

    return total_chars;
}

File: shell.php
<?php

pcntl_async_signals(true);
function handler($signal) {
    echo "caught signal $signal\n";
}
pcntl_signal(SIGINT, "handler");
pcntl_signal(SIGHUP, "handler");
pcntl_signal(SIGUSR1, "handler");
pcntl_signal(SIGTERM, "handler");

$start = microtime(true);
$output = "";
$cmd = "./new";

$i = 0;
while (microtime(true) - $start < 1) {
    list($stdout) = shellout("./new");
    $output .= $stdout;
    $i++;
}

$time = microtime(true) - $start;
printf("PHP SHELL: Ran %d times in %f seconds.\n", $i, $time);

function shellout(string $cmd): array {
    $descriptors = [
        0 => ['pipe', 'r'],  // stdin
        1 => ['pipe', 'w'],  // stdout
        2 => ['pipe', 'w']   // stderr
    ];
    $fd = proc_open("./new", $descriptors, $pipes);
    if (!is_resource($fd)) {
        return ["", "command '$cmd' does not exist", 127];
    }

    $stdout = stream_get_contents($pipes[1]);
    fclose($pipes[1]);

    $stderr = stream_get_contents($pipes[2]);
    fclose($pipes[2]);

    $code = proc_close($fd);

    return [$stdout, $stderr, $code];
}
